### PR TITLE
Search: add cache tags

### PR DESCRIPTION
**Pull request by stsewd**
So, there is one catch, search from the main project should be purged when one of the versions from its subprojects change (this can be done by adding one extra query when purging versions).

And... we are also losing some data from search analytics here, so not really sure if we want to cache search after all, or maybe just cache for a shorter time.

### CODE REVIEW

**Comment by ericholscher** [comment]
I think the UX benefits of search caching are definitely better than search analytics being exact. The numbers will still be approximately correct, and if we care a lot we could send the search analytics data to an endpoint via JS, but I don't think we do :)

**Comment by stsewd** [comment]
ok, then I'll figure out the subprojects problem, probably it will end up like the footer having an extra tag `{project.slug}-rtd-search`

**Review by ericholscher — APPROVED** [review]
> So, there is one catch, search from the main project should be purged when one of the versions from its subprojects change (this can be done by adding one extra query when purging versions).

Hrm, I feel like we keep hitting this issue :) Similar to the footer cache purging as well, I guess? I wonder if we can combine them into a larger framework for thinking about this.

File: readthedocs/api/v2/mixins.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    Add cache tags for project and version to the response of this view.
+
+    The view inheriting this mixin should implement the
+    `self._get_project` and `self._get_version` methods.
+    """
+
+    def dispatch(self, request, *args, **kwargs):
+        response = super().dispatch(request, *args, **kwargs)
+        cache_tags = self._get_cache_tags()
+        response['Cache-Tag'] = ','.join(cache_tags)
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/api/v2/mixins.py:18]: Not sure if it's ok, but better to be sure p:

**Comment by stsewd** [comment]
> Hrm, I feel like we keep hitting this issue :) Similar to the footer cache purging as well, I guess?

More or less, for footer we don't need to purge parent projects, but the footer from all versions. For search, we don't need to purge the search from all versions, but from the main project.

**Pull request merged**